runtime/buffer.h

```diff
@@ -334,85 +335,85 @@ inline bool can_fuse(const dim& inner, const dim& outer) {
-    // already points to the min.
+    // already points to begin.
     for_each_contiguous_slice(base, dims, d - 1, elem_size, f, slice_extent);
     for (index_t i = begin + 1; i < end; ++i) {
-      for_each_contiguous_slice(
-          offset_bytes(base, dim.flat_offset_bytes(i)), dims, d - 1, elem_size, f, slice_extent);
+      for_each_contiguous_slice(offset_bytes(base, dim.flat_offset_bytes(i)), dims, d - 1, elem_size, f, slice_extent);
     }
   }
 }
 
 // Implements the cropping part of a loop over tiles.
-template <int D, typename Tile, typename F>
-void for_each_tile(const Tile& tile, raw_buffer& buf, F&& f) {
-  if constexpr (D == -1) {
+template <typename F>
+void for_each_tile(const index_t* tile, raw_buffer& buf, int d, const F& f) {
+  if (d == -1) {
     f(buf);
-  } else if constexpr (std::is_same_v<std::tuple_element<D, Tile>, all>) {
-    // Don't want to tile this dimension.
-    for_each_tile<D - 1>(tile, buf, f);
+    return;
+  }
+
+  slinky::dim& dim = buf.dim(d);
+  index_t step = tile[d];
+  if (dim.extent() <= step) {
+    // Don't need to tile this dimension.
+    for_each_tile(tile, buf, d - 1, f);
   } else {
-    slinky::dim& dim = buf.dim(D);
+    // TODO: Supporting folding here should be possible.
+    assert(dim.fold_factor() == dim::unfolded);
+    index_t stride = dim.stride() * step;
 
-    // Get the tile size in this dimension.
-    index_t step = std::get<D>(tile);
-    if (dim.extent() <= step) {
-      // There's nothing to crop here, just move on.
-      for_each_tile<D - 1>(tile, buf, f);
-    } else {
-      // TODO: Supporting folding here should be possible.
-      assert(dim.fold_factor() == dim::unfolded);
-      index_t stride = dim.stride() * step;
-
-      // Save the old base and bounds.
-      void* old_base = buf.base;
-      index_t old_min = dim.min();
-      index_t old_max = dim.max();
-
-      // Handle the first tile.
-      dim.set_bounds(old_min, std::min(old_max, old_min + step - 1));
-      for_each_tile<D - 1>(tile, buf, f);
-      for (index_t i = old_min + step; i <= old_max; i += step) {
-        buf.base = offset_bytes(buf.base, stride);
-        dim.set_bounds(i, std::min(old_max, i + step - 1));
-        for_each_tile<D - 1>(tile, buf, f);
-      }
-
-      // Restore the old base and bounds.
-      buf.base = old_base;
-      dim.set_bounds(old_min, old_max);
+    // Save the old base and bounds.
+    void* old_base = buf.base;
+    index_t old_min = dim.min();
+    index_t old_max = dim.max();
+
+    // Handle the first tile.
+    dim.set_bounds(old_min, old_min + step - 1);
+    for_each_tile(tile, buf, d - 1, f);
+    for (index_t i = old_min + step; i <= old_max; i += step) {
+      buf.base = offset_bytes(buf.base, stride);
+      dim.set_bounds(i, std::min(old_max, i + step - 1));
+      for_each_tile(tile, buf, d - 1, f);
     }
+
+    // Restore the old base and bounds.
```

Comment:
Mutating the buffer in-place seems like a future bug waiting to happen, if/when we get async code plugged in. At a minimum, maybe document this function as not thread-safe.